File: project_euler_10.py
# The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.
# Find the sum of all the primes below two million.

import time
def is_prime(number, primes_list):
    """
    Checks if a certain number is a prime number using a list of prime numbers
    as auxiliary so that less calculation is required
    :param number: number that will be checked
    :param primes_list: list with all prime numbers smaller than number
    :return: true if prime, false if not
    """
    if (number % 10 == 5):
        return False

    if primes_list[round(primes_list * 9 / 16 - 1)] > number // 2:
        primes_list = primes_list[0:round(primes_list * 9 / 16 - 1)]
    elif primes_list[round(primes_list * 5 / 8 - 1)] > number // 2:
        primes_list = primes_list[0:round(primes_list * 5 / 8 - 1)]
    elif primes_list[round(primes_list * 3 / 4 - 1)] > number // 2:
        primes_list = primes_list[0:round(primes_list * 3 / 4 - 1)]
    for prime in primes_list:
            if number % prime == 0:
                return False
    return True

start = time.time()
primes_list = [2, 3, 5] # will use this to reduce amount of numbers to test in the
Sum = 10  # starting with number 10 so the 'for' below can skip 2,3,5
for i in range(7, 2000000, 2):
    if is_prime(i, primes_list):
        Sum += i
        primes_list.append(i)

end = time.time()
print(Sum, end - start)

File: project_euler_10.c
#include <time.h>

int v[1000000]; // vector that we will add all prime numbers found
int array_index = 0; // index that contains the latest prime added to v
int last_prime = 0; // index that contains the last prime in v that is imeddiately
//bigger than half of the next number to be checked as prime to reduce amount
//of testing in is_prime function

int main(){ 
     time_t start;
     time_t end;
     start = time(NULL);
     float sum = 2; // start from 2 so that we can skip even numbers without
     //adding another if statement in our for loop
     v[0] = 2;
     int i;
     for(i = 3; i<2000000 ;i = i + 2){ //skipping all even numbers, because
     //they are not prime
           if(is_prime(i)){
                           sum += i;
                           v[array_index] = i;
                           while(v[last_prime] < (i+2)/2){
                                   last_prime++;        
                           }
                           array_index++;
           }
     }

     end = time(NULL); 
     printf("Runtime %ld\n", end - start);
     printf("Sum: %f", sum);
     getchar();
     return(0);
       
}
          

int is_prime(int number){
        int i;
        if (number != 5 && number % 10 == 5){
                  return(0);          
        }
        for( i = 0; i <= last_prime; i++){
              if(number % v[i] == 0){
                        return(0); 
              }      
        }        
        return(1);
        
        
}
